### ISSUE TITLE
Previously focused app takes focus while the currently focused app is grabbed

### ISSUE THREAD

**Issue by git-f0x** [pop-os/cosmic-comp#453]
The active hint doesn't move, but the previous app becomes visibly focused when the current app is grabbed.
When e.g. the search bar in Firefox is selected, and then switching to another app and grabbing it, typing while the window is grabbed will input text to the Firefox search bar.

**Comment by Drakulix** [comment]
fixed by https://github.com/pop-os/cosmic-comp/commit/d4926c3a9e66a2b699fbe626b2609ed5d373cce4

**Comment by PoisonFrog** [comment]
I'm still having this issue on master.

**Comment by Drakulix** [comment]
> I'm still having this issue on master.

Which distro? Which commit exactly?

**Comment by PoisonFrog** [comment]
> > I'm still having this issue on master.
> 
> Which distro? Which commit exactly?

pop_OS!
I added popdev:master to the sources and the system is fully up to date

**Comment by PoisonFrog** [comment]
I just realized that it doesn't happen when you drag the title bar, but it still happens when you drag the window while holding the Super key.

**Comment by git-f0x** [comment]
I can kind of reproduce it if the previously focused app is a GTK app (that app visually changes when another is dragged), but doesn't have any usability issues (e.g. text input). It's just a visual issue.

**Comment by PoisonFrog** [comment]
> I can kind of reproduce it if the previously focused app is a GTK app (that app visually changes when another is dragged), but doesn't have any usability issues (e.g. text input). It's just a visual issue.

I still have this issue with all kinds of apps, including libcosmic apps and I can still input to the background window if holding the Super key to drag. If dragging the title bar I only have this visual issue with GTK apps.

**Comment by git-f0x** [comment]
I can reproduce the visual changes in most apps with Super dragging, but text input doesn't go through.

**Comment by Drakulix** [comment]
> I just realized that it doesn't happen when you drag the title bar, but it still happens when you drag the window while holding the Super key.

Oh thanks, that is what I was missing here. Reopening.

**Comment by Drakulix** [comment]
Fixed by https://github.com/pop-os/cosmic-comp/commit/6c517bd8de9e4de74a24866c274b2e1c8ac6d7e1

**Comment by PoisonFrog** [comment]
It's mostly fixed now, but there are still a few quirks. For example, if you drag a window (using either the title bar or the Super key) and press Super + Q, it closes the window in the background instead of the one you're dragging. This also happens with other window shortcuts like maximize.

I also noticed a minor bug(?) with input while dragging. When you long-press a key in an input field it behaves differently if you're dragging the window. Normally long key presses make the input repeat, but if you're dragging it doesn't unless you press it for long enough then the input repeats for a second or two and then it stops again. It's a bit weird and I'm not sure if this is intended or not. It's not anything major and should not affect users negatively anyways.

**Comment by git-f0x** [comment]
Some visual issues that might point to some unexpected behavior:
The previously focused GTK app or an app using SSDs changes to visually focused when the currently focused app is dragged (Qt doesn't seem to be affected).
Regardless of the toolkit/app, the App Tray highlight switches to the previously focused app during the drag.

**Comment by PoisonFrog** [comment]
Another bug:

If I _SUPER + Left Click_ some apps to drag them they will register the left click input. This doesn't affect libcosmic apps and I don't think it affects libadwaita apps as well, but it does affect kitty terminal and the non-libadwaita version of Nautilus of Pop!_OS 22.04. Probably other apps too.

**Comment by git-f0x** [comment]
That's a different issue, and not related to this. It would probably be better to add that comment to #544.

**Comment by RedBearAK** [comment]
@Drakulix @WatchMkr 

I just confirmed that on an updated install of Pop 24.04 LTS (and Fedora 41) in a virtual machine, this issue of receiving "activated" events for the window just behind the front window is still there. I know for a fact that the Fedora 41 install has COSMIC alpha 4, and I presume that the updated Pop 24.04 virtual machine also has package versions from alpha 4 at this point. But the package versioning on Pop is not as straightforward. 

Having thought about this a lot, it seems very probable that this issue will end up being a simple off-by-one error in the code the emits the Wayland state events when the window is clicked and dragged. It's like the code sends out the event for the window at position one/"1" in the window stack, instead of sending out the event for the window at position zero/"0" in the window stack. Thus the window behind the actual focused window is getting marked as "active" from that point until another window state event occurs to correct the error. 

This issue will present a problem for any Linux keymapper that offers app-specific remapping on COSMIC. And for any other app that needs to know which window is currently active. Any dragged window will push the "active" info out of alignment with reality until some other window switch or maximize/restore event happens to reset that info with another "activated" state event. 

Also the empty Wayland state events and unexplained all-zeroes Wayland state events are still there. All as described previously. 

In other words, nothing has noticeably changed about this issue through multiple alpha releases, and the beta is supposedly coming relatively soon. Just wondering if anyone is paying any attention to this. I know there are a lot of other things being worked on, but I feel like accurately advertising which window is active is a somewhat fundamental task for a window manager.

**Comment by PoisonFrog** [comment]
I just realized that if you open a window in your last workspace so a new workspace is created dynamically, and then you drag the window you just opened, then the dynamically created workspace that was just added will disappear. Once you release the window it will be created again. At least that's what the workspaces applet reports. I think this is a related issue.

My guess is that while you are dragging the window the compositor behaves like that window doesn't exist anymore for some stuff, and that's why focus goes to the window behind or the adjacent dynamically created workspace disappears if the dragged window is the only one in the current workspace.

**Comment by ids1024** [comment]
It looks like in https://github.com/pop-os/cosmic-comp/blob/master/src/input/actions.rs, `Action::Close` operates on the keyboard target (which may be a window being dragged, or a focused window on an output other than the one with the cursor), while I lot of other bindings use the focus stack on the current output.

Presumably we at least want all the bindings operating on the "current window" to be consistent in how they define that? Using the same surface that receives keyboard input seems intuitively expected.

> My guess is that while you are dragging the window the compositor behaves like that window doesn't exist anymore for some stuff

I guess you could say the window is treated as not existing on any workspace while it is being dragged.

**Comment by git-f0x** [comment]
I think this can be closed. :)
